src/KDMqtt/mqtt.cpp

```diff
@@ -0,0 +1,523 @@
+}
+
+void MqttClient::onWriteOpRequested()
+{
+    const auto writeOpIsPending = m_mosquitto.client()->wantWrite();
+    if (!writeOpIsPending) {
+        return;
+    }
+
+    auto result = m_mosquitto.client()->loopWrite();
+    MqttLib::instance().checkMosquittoResultAndDoDebugPrints(result, "loopWrite()");
+}
+
+void MqttClient::onMiscTaskRequested()
+{
+    auto result = m_mosquitto.client()->loopMisc();
+    MqttLib::instance().checkMosquittoResultAndDoDebugPrints(result, "loopMisc()");
+}
+
+void MqttClient::EventLoopHook::init(const std::chrono::milliseconds miscTaskInterval, MqttClient *parent)
+{
+    spdlog::debug("MqttClient::EventLoopHook::init()");
+    assert(parent != nullptr);
+
+    this->parent = parent;
+
+    miscTaskTimer = std::make_unique<Timer>();
+    miscTaskTimer->interval.set(miscTaskInterval);
+    miscTaskTimer->running.set(false);
+    std::ignore = miscTaskTimer->timeout.connect(&MqttClient::onMiscTaskRequested, parent);
+}
+
+void MqttClient::EventLoopHook::engage(const int socket)
+{
+    spdlog::debug("MqttClient::EventLoopHook::engage()");
+
+    if (!isSetup()) {
+        spdlog::error("MqttClient::EventLoopHook::engage() - EventLoopHook is not initialized. Call MqttClient::EventLoopHook::init() first.");
+        return;
+    }
+
+    if (isEngaged()) {
+        spdlog::error("MqttClient::EventLoopHook::engage() - Already engaged.");
+        return;
+    }
+
+    if (socket < 0) {
+        spdlog::error("MqttClient::EventLoopHook::engage() - Invalid socket.");
+        return;
+    }
+
+    readOpNotifier = std::make_unique<FileDescriptorNotifier>(socket, FileDescriptorNotifier::NotificationType::Read);
+    writeOpNotifier = std::make_unique<FileDescriptorNotifier>(socket, FileDescriptorNotifier::NotificationType::Write);
+
+    std::ignore = readOpNotifier->triggered.connect(&MqttClient::onReadOpRequested, parent);
+    std::ignore = writeOpNotifier->triggered.connect(&MqttClient::onWriteOpRequested, parent);
+
+    miscTaskTimer->running.set(true);
+}
+
+void MqttClient::EventLoopHook::disengage()
+{
+    spdlog::debug("MqttClient::EventLoopHook::disengage()");
+
+    if (!isEngaged()) {
+        spdlog::error("MqttClient::EventLoopHook::disengage() - Already disengaged.");
+        return;
+    }
+
+    miscTaskTimer->running.set(false);
+
+    readOpNotifier->triggered.disconnectAll();
+    writeOpNotifier->triggered.disconnectAll();
+
+    readOpNotifier = {};
+    writeOpNotifier = {};
+}
+
+bool MqttClient::EventLoopHook::isSetup() const
+{
+    return (miscTaskTimer && (parent != nullptr));
+}
+
+bool MqttClient::EventLoopHook::isEngaged() const
+{
+    return (readOpNotifier && writeOpNotifier);
+}
+
+void MqttClient::MosquittoClientDependency::init(MosquittoClient *client, MqttClient *parent)
+{
+    spdlog::debug("MqttClient::MosquittoClientDependency::init()");
+    assert(parent != nullptr);
+
+    delete mosquittoClient;
+    mosquittoClient = client;
```

Comment:
`MosquittoClientDependency` takes ownership of `MosquittoClient` pointer here. We delete the old `mosquittoClient` (if any) but It seems we don't delete it on destruction and leak.